src/cli/cli_dns.cpp

```diff
@@ -650,6 +699,78 @@ void Dns::HandleDnsServiceResponse(otError aError, const otDnsServiceResponse *a
 }
 
 #endif // OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+
+#if OPENTHREAD_CONFIG_DNS_CLIENT_ARBITRARY_RECORD_QUERY_ENABLE
+void Dns::HandleDnsRecordResponse(otError aError, const otDnsRecordResponse *aResponse, void *aContext)
+{
+    static_cast<Dns *>(aContext)->HandleDnsRecordResponse(aError, aResponse);
+}
+
+void Dns::HandleDnsRecordResponse(otError aError, const otDnsRecordResponse *aResponse)
+{
+    char            name[OT_DNS_MAX_NAME_SIZE];
+    uint8_t         data[kMaxRrDataSize];
+    otDnsRecordInfo recordInfo;
+
+    IgnoreError(otDnsRecordResponseGetQueryName(aResponse, name, sizeof(name)));
+
+    OutputLine("DNS query response for %s ", name);
+
+    SuccessOrExit(aError);
+
+    for (uint16_t index = 0;; index++)
+    {
+        ClearAllBytes(recordInfo);
+        recordInfo.mNameBuffer     = name;
+        recordInfo.mNameBufferSize = sizeof(name);
+        recordInfo.mDataBuffer     = data;
+        recordInfo.mDataBufferSize = sizeof(name);
+
+        aError = otDnsRecordResponseGetRecordInfo(aResponse, index, &recordInfo);
+
+        if (aError == OT_ERROR_NOT_FOUND)
+        {
+            aError = OT_ERROR_NONE;
+            ExitNow();
+        }
+
+        SuccessOrExit(aError);
+
+        OutputLine("%u)", index);
+        OutputLine(kIndentSize, "RecordType:%u, RecordLength:%u, TTL:%lu, Section:%s", recordInfo.mRecordType,
+                   recordInfo.mRecordLength, ToUlong(recordInfo.mTtl), RecordSectionToString(recordInfo.mSection));
+        OutputLine(kIndentSize, "Name:%s", recordInfo.mNameBuffer);
+        OutputFormat(kIndentSize, "RecordData:[");
+        OutputBytes(recordInfo.mDataBuffer, recordInfo.mDataBufferSize);
+
+        if (recordInfo.mDataBufferSize != recordInfo.mRecordLength)
+        {
+            OutputFormat("...");
+        }
+
+        OutputLine("]");
+    }
+
+exit:
+    OutputResult(aError);
+}
+
+const char *Dns::RecordSectionToString(otDnsRecordSection aSection)
+{
+    const char *const kSectionString[] = {
+        "answer",          // (0) OT_DNS_SECTION_ANSWER
+        "authority",       // (1) OT_DNS_SECTION_AUTHORITY
+        "additional data", // (2) OT_DNS_SECTION_ADDITIONAL_DATA
```

Comment:
RFC 1035 uses "additional section" or "additional records" section. Follow-up DNS RFCs use one of these or "additional data" section. So it's very inconsistent overall. Given that, I would use only "additional" here.

```suggestion
        "additional", // (2) OT_DNS_SECTION_ADDITIONAL_DATA
```

Reply:
Also remove the `_DATA` from the Additional section name, I think. (For consistency - it should be clear enough.)